\documentclass[11pt]{article}

\title{Enabling high throughput haplotype analysis through hardware acceleration}
\author{Carneiro MO, Biagioli E, Pratt-Szeliga P, Thibault S, Vacek G, Nehab D}

\begin{document}
	\maketitle
	
	\begin{abstract}
	Small variants such as single nucleotide polymorphisms (SNPs) or single base pair insertions or deletions (indel) are not the most difficult task for variant calling algorithms. Latest callsets from the 1000 Genomes project have shown upwards of 99\% sensitivity on those types of variants (cite 1kg). Calling larger indels and short rearrangements (up to 50 bases long) can only be achieved through local denovo assembly of a reference haplotype based on the read data (cite haplotype caller?). For this computation we rely on fast heuristic methods that generate pcandidate haplotypes without trying all possibilties (cite debruijn graphs). Once we have a list of candidate haplotypes, we have to determine which of these haplotypes is the most likely haplotype to explain the read data. For this task we use a paired hidden markov model (PairHMM) that calculates the probabiltity of all alignments of a read to a candidate haplotype, repeating the process for every candidate haplotype against every read and tallying the results. The haplotype with the highest probability across all reads is the chosen one. This PairHMM is responsible for more than 50\% of the runtime of the Haplotype Caller (denovo assembly variant caller available in the Genome Analysis Toolkit) due to it's $N^2$ implementation. Here we show how we can harness the parallelism of graphics processing units (GPU) and field-programmable gate arrays (FPGA) to reduce the runtime by ~300 fold. This performance improvement not only accelerates the calling process, but also enables projects that need to call tens of thousands of samples which would otherwise be impossible.
	\end{abstract}

\end{document}

